### PR TITLE
Support Cluster Management

### PR DESCRIPTION
**Pull request by darionyaphet**
#### What type of PR is this?
- [ ] bug
- [X] feature
- [ ] enhancement

#### Checklist：
- [ ] Documentation affected （Please add the label if documentation needs to be modified.)
- [ ] Incompatible （If it is incompatible, please describe it and add corresponding label.）
- [ ] Need to cherry-pick （If need to cherry-pick to some branches, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes：
Please confirm whether to reflect in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/meta/processors/zone/DropHostsProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    iter->next();
+  }
+  return nebula::cpp2::ErrorCode::SUCCEEDED;
+}
+
+void DropHostsProcessor::rewriteSpaceProperties(const std::string& spaceKey,
+                                                const meta::cpp2::SpaceDesc& properties) {
+  auto spaceVal = MetaKeyUtils::spaceVal(properties);
+  std::vector<kvstore::KV> data = {{spaceKey, spaceVal}};
+  doSyncPutAndUpdate(std::move(data));
```
Comment (critical27) [inline src/meta/processors/zone/DropHostsProcessor.cpp:152]: Don't write directly, you should make sure all writes succeed or failed.

File: src/meta/processors/zone/RenameZoneProcessor.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * This source code is licensed under Apache 2.0 License.
+ */
+
+#include "meta/processors/zone/RenameZoneProcessor.h"
+
+namespace nebula {
+namespace meta {
+
+void RenameZoneProcessor::process(const cpp2::RenameZoneReq& req) {
```
Comment (critical27) [inline src/meta/processors/zone/RenameZoneProcessor.cpp:11]: Lock is missing

File: src/meta/processors/zone/DropZoneProcessor.cpp [hunk 3]
```diff
@@ -24,43 +24,12 @@ void DropZoneProcessor::process(const cpp2::DropZoneReq& req) {
     return;
   }
```
Comment (critical27) [inline src/meta/processors/zone/DropZoneProcessor.cpp:26]: Where do we check host is empty? The space properties is not impacted?

Comment (darionyaphet) [reply src/meta/processors/zone/DropZoneProcessor.cpp:26]: `DROP ZONE` will be fixed in the next pr. This one only remove some group related function

File: src/meta/processors/zone/DropHostsProcessor.cpp (hunk 1 above)
Comment (darionyaphet) [reply src/meta/processors/zone/DropHostsProcessor.cpp:152]: fine

File: src/meta/ActiveHostsMan.cpp [hunk 4]
```diff
@@ -145,7 +177,7 @@ ErrorOr<nebula::cpp2::ErrorCode, std::vector<HostAddr>> ActiveHostsMan::getActiv
   return activeHosts;
 }
 
-ErrorOr<nebula::cpp2::ErrorCode, std::vector<HostAddr>> ActiveHostsMan::getActiveHostsWithGroup(
+ErrorOr<nebula::cpp2::ErrorCode, std::vector<HostAddr>> ActiveHostsMan::getActiveHostsWithZones(
```
[the remark this answers is not in the document]
Comment (liwenhui-soul) [reply src/meta/ActiveHostsMan.cpp:179]: > 
sure,I don't use it any more

**Comment by liwenhui-soul** [comment]
seems like CREATE ZONE is missing?

File: src/meta/processors/zone/AddHostsProcessor.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    // Ensure that the node is not registered.
+    auto machineKey = MetaKeyUtils::machineKey(host.host, host.port);
+    if (machineExist(machineKey) == nebula::cpp2::ErrorCode::SUCCEEDED) {
+      LOG(ERROR) << "The host " << host << " have existed!";
+      code = nebula::cpp2::ErrorCode::E_EXISTED;
+      break;
+    }
+
+    // Automatic generation zone
+    auto zoneName = folly::stringPrintf("default_zone_%s_%d", host.host.c_str(), host.port);
```
Comment (liwenhui-soul) [inline src/meta/processors/zone/AddHostsProcessor.cpp:45]: I got this error
(root@nebula) [(none)]> rename zone default_zone_127.0.0.1_40180 to z2
[ERROR (-1004)]: SyntaxError: syntax error near `.0'

I think the zone_name should not contain "." .

File: src/parser/parser.yy [hunk 6]
```diff
@@ -3289,14 +3302,14 @@ create_space_sentence
         sentence->setComment($5);
         $$ = sentence;
     }
-    | KW_CREATE KW_SPACE opt_if_not_exists name_label KW_ON name_label {
+    | KW_CREATE KW_SPACE opt_if_not_exists name_label KW_ON zone_name_list {
```
Comment (liwenhui-soul) [inline src/parser/parser.yy:3330]: missing pattern that create space without zone_name_list? I see your code implementation that create space without zone_name_list means create space on all zones.

**Comment by liwenhui-soul** [comment]
LGTM

**Comment by liwenhui-soul** [comment]
LGTM

File: src/meta/ActiveHostsMan.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (critical27) [reply src/meta/ActiveHostsMan.cpp:179]: @darionyaphet plz delete all related interfaces

File: src/meta/processors/zone/DropHostsProcessor.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    LOG(ERROR) << "List zones failed, error: " << apache::thrift::util::enumNameSafe(retCode);
+    handleErrorCode(retCode);
+    onFinished();
+    return;
+  }
+
+  auto iter = nebula::value(iterRet).get();
+  while (iter->valid()) {
+    auto zoneKey = iter->key();
+    auto hs = MetaKeyUtils::parseZoneHosts(iter->val());
```
Comment (critical27) [inline src/meta/processors/zone/DropHostsProcessor.cpp:84]: 1. The if branch in line 86
    * Where do we check each host in `hs` has no related partitions?
    * If all hosts has no related partitions, the zone itself should be deleted
2. The else branch in 101
    * Where do we check each host in `hs` has no related partitions?

**Review by critical27 — APPROVED** [review]
LGTM now, good job

**Review by liwenhui-soul — DISMISSED** [review]

**Comment by critical27** [comment]
Please write some document about what should user to do after the PR merged.

File: src/graph/validator/MaintainValidator.cpp [hunk 8]
```diff
@@ -460,11 +460,23 @@ Status ShowEdgeIndexStatusValidator::toPlan() {
...
-Status AddZoneValidator::toPlan() {
-  auto sentence = static_cast<AddZoneSentence *>(sentence_);
-  auto *doNode = AddZone::make(qctx_, nullptr, *sentence->zoneName(), sentence->hosts()->hosts());
+Status MergeZoneValidator::toPlan() {
+  auto sentence = static_cast<MergeZoneSentence *>(sentence_);
+  auto *doNode =
+      MergeZone::make(qctx_, nullptr, *sentence->zoneName(), sentence->zoneNames()->zoneNames());
+  root_ = doNode;
+  tail_ = root_;
+  return Status::OK();
```
[the remark this answers is not in the document]
Comment (darionyaphet) [reply src/graph/validator/MaintainValidator.cpp:471]: merge and split will support in another PR

File: src/clients/meta/MetaClient.cpp [hunk 9]
```diff
@@ -3070,33 +3118,35 @@ folly::Future<StatusOr<bool>> MetaClient::dropZone(std::string zoneName) {
   return future;
 }
 
-folly::Future<StatusOr<bool>> MetaClient::addHostIntoZone(HostAddr node, std::string zoneName) {
-  cpp2::AddHostIntoZoneReq req;
-  req.set_node(node);
+folly::Future<StatusOr<bool>> MetaClient::splitZone(
+    std::string zoneName, std::unordered_map<std::string, std::vector<HostAddr>>) {
```
[the remark this answers is not in the document]
Comment (darionyaphet) [reply src/clients/meta/MetaClient.cpp:3122]: fix in  another PR

File: src/graph/executor/admin/SpaceExecutor.cpp [hunk 10]
```diff
@@ -226,31 +225,30 @@ folly::Future<Status> ShowCreateSpaceExecutor::execute() {
...
+              folly::stringPrintf(fmt,
+                                  properties.get_space_name().c_str(),
+                                  properties.get_partition_num(),
+                                  properties.get_replica_factor(),
+                                  properties.get_charset_name().c_str(),
+                                  properties.get_collate_name().c_str(),
+                                  SchemaUtil::typeToString(properties.get_vid_type()).c_str(),
+                                  sAtomicEdge.c_str(),
+                                  zoneNames.c_str(),
+                                  properties.comment_ref()->c_str()));
```
[the remark this answers is not in the document]
Comment (darionyaphet) [reply src/graph/executor/admin/SpaceExecutor.cpp:240]: It doesn't look very good because it's in the middle

**Comment by wenhaocs** [comment]
Otherwise, LGTM and thank you for enhancing the code out of the zoning project in this PR.

**Review by wenhaocs — APPROVED** [review]

**Review by liwenhui-soul — APPROVED** [review]

**Pull request merged**